content/en/blog/2022/rust-instrumentation/index.md

```diff
@@ -0,0 +1,396 @@
+
+    let new_user = models::User {
+        id: Uuid::new_v4().to_string(),
+        name: nm.to_owned(),
+    };
+    diesel::insert_into(users).values(&new_user).execute(conn)?;
+
+    Ok(new_user)
+}
+```
+
+And for the diesel library which we are using. To add the diesel instrumentation we will need to use diesel_tracing crate and replace our PgConnection with InstrumentedPgConnection:
+
+```Rust
+use diesel_tracing::pg::InstrumentedPgConnection;
+
+pub type DbConnection = InstrumentedPgConnection;
+
+pub type DbPool = r2d2::Pool<ConnectionManager<DbConnection>>;
+```
+
+I ran the application once more and observed the traces in Jaeger:
+
+![Viewing traces in Jaeger](traces-in-jaeger.png)
+
+We can see many unrelated traces. This is because of two reasons.
+
+The first reason is that the database connection struct communicates with the database periodically by calling the establish function to make sure the connection is still live. The second reason is explained below.
+
+### How to preserve our span context when executing closures on different threads?
+
+But there is a problem here. Can you see it? For some reason, the spans representing our request and the spans representing the database query are not grouped together. This is happening because of this code:
+
+```Rust
+web::block(move || {
+        let conn = db.get()?;
+        db_operations::insert_new_user(&new_user.name, &conn)
+    }).await?
+```
+
+We are invoking our db operation inside a closure we pass to the web::block function. We lose the span context of our execution flow. However, the web::block function is crucial.
+
+It executes a blocking function on a thread pool dedicated to async blocking operations. So how can we keep the context of the trace?
+
+We can do that by wrapping the web::block function with our own tracing function:
+
+```Rust
+fn traced_web_block<F, R>(f: F) -> impl Future<Output = Result<R, BlockingError>>
+where
+    F: FnOnce() -> R + Send + 'static,
+    R: Send + 'static,
+{
+    let current_span = tracing::Span::current();
+    web::block(move || current_span.in_scope(f))
+}
+```
+
+Here we get our current span and invoking the closure within its scope.
+
+Let’s view our traces now:
+
+![View a trace in Jaeger](trace-view.png)
+
+We can now get a full picture of the flow of our application.
+
+## Conclusion
+
+We reviewed how to add distributed tracing with OpenTelemetry to the Rust application. We explored how to add both manual instrumentation and library-provided instrumentations.
+
+We used Jaeger to visualize the flow of our system and learned what to do when the flow is broken.
+
+To sum up, I enjoyed working on this example. I think the approach Rust is taking regarding tracing is interesting and worth considering when writing a Rust program. Furthermore, the ecosystem is growing so I am sure more libraries will provide instrumentations soon.
+
+I hope you learned something new. Feel free to reach out with any questions you might have.
```

Comment:
> Feel free to reach out ...

Let folks know how to reach out, e.g.:  Slack, GitHub discussions or issues. And provide link(s).